Sources/HTTP/HTTPConnection.swift

```diff
@@ -0,0 +1,620 @@
+  }
+  
+  private final func processDataCB(newState s: LastCallback,
+                                   p: UnsafePointer<CChar>?, len: size_t)
+                     -> Int32
+  {
+    let newState = s
+    if lastCallBack == newState { // continue value
+      if let p = p {
+        let bp = UnsafeBufferPointer(start: p, count: len)
+        parserBuffer.append(bp)
+      }
+      return 0 // done already. state is the same
+    }
+    
+    switch lastCallBack { // != newState!
+      case .url: // finished URL
+        if !parserBuffer.isEmpty {
+          parsedURL = String(data: parserBuffer, encoding: .utf8)
+        }
+      
+      case .field: // last field was a name
+        lastHeaderName = headerNameForData(parserBuffer)
+      
+      case .value: // last field was a value, now something new
+        let value = String(data: parserBuffer, encoding: .utf8)
+        #if DEBUG
+          assert(lastHeaderName != nil, "header value w/o a name?")
+          assert(value          != nil, "header value missing?")
+        #endif
+        
+        // TBD: Uh, oh, why the need to create a literal here??
+        if let name = lastHeaderName, let value = value {
+          parsedHeaders.append([name: value])
+        }
+        lastHeaderName = nil
+      
+      default:
+        break
+    }
+    
+    // store new data & state
+    parserBuffer.removeAll()
+    lastCallBack = newState
+    if len > 0, let p = p {
+      let bp = UnsafeBufferPointer(start: p, count: len)
+      parserBuffer.append(bp)
+    }
+    return 0
+  }
+  
+  @inline(__always)
+  static func getSelf(parser: UnsafeMutablePointer<http_parser>?)
+              -> HTTPConnection?
+  {
+    guard let pointee = parser?.pointee.data else { return nil }
+    return Unmanaged<HTTPConnection>
+             .fromOpaque(pointee)
+             .takeUnretainedValue()
+  }
+  
+  
+  // MARK: - CustomStringConvertible
+  
+  var description : String {
+    var ms = "<HTTPConnection:"
+    
+    ms += " fd=\(channel.fileDescriptor)"
+    
+    if !isReadSideOpen {
+      ms += " read-closed"
+    }
+    
+    if let readBuffer = readBuffer {
+      ms += " buffered=#\(readBuffer.count)"
+      if bufferedEOF { ms += "/+EOF" }
+    }
+    else if bufferedEOF {
+      ms += " buffered-EOF"
+    }
+    
+    if bodyHandler != nil { ms += " body" }
+    
+    if writers.isEmpty {
+      ms += " no-writers"
+    }
+    else if writers.count == 1 {
+      ms += " writer=\(writers[0])"
+    }
+    else {
+      ms += " headwriter=\(writers[0]) of #\(writers.count)"
+    }
+    
+    ms += ">"
+    return ms
+  }
+}
+
+
+// MARK: - Static Strings
+
+@inline(__always)
+fileprivate func headerNameForData(_ data: Data) -> HTTPHeaders.Name? {
+  guard !data.isEmpty else { return nil }
+  // TODO: reuse header values by matching the data via memcmp(), maybe first
+  //       switch on length, compare c0
+  guard let s = String(data: data, encoding: .utf8) else { return nil }
```

Comment:
Can put these both in one `guard` if you like

Reply:
Why would I do this? Those are separate things. This one doesn't make sense for me.